File: code/cpp/usrIpDFSALR.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include<map>

// data structure to store an edge
struct Edge
{
    int src, dest, cost;

    // constructor
    Edge(int s, int d, int c=1)
    {
        src = s;
        dest = d;
        cost = c;
    } 
};

class Graph
{
public:
    int numNodes;
    std::map<int, std::vector<std::pair<int, int>>> adjList;

    // constrcutor
    Graph(int numNodes, std::vector<Edge> const &edges, bool directed=false)
    {
        this->numNodes = numNodes;

        for (auto &edge : edges)
        {
            adjList[edge.src].push_back({edge.dest, edge.cost});
            
            if (!directed)
                adjList[edge.dest].push_back({edge.src, edge.cost});
        }       
    }
};

void dfs(Graph &graph, std::vector<bool> &visited, int v=0)
{
    // mark node as visited and print its value
    visited[v] = true;
    std::cout << v << " ";

    // for each neighbour in adj. List of v call dfs
    // std::pair<int, int> instead of auto 
    for (auto u : graph.adjList[v])
    {
        if (!visited[u.first])
            dfs(graph, visited, u.first);
    }
}

long dfsCountNode(Graph &graph, std::vector<bool> &visited, int v=0)
{
    if (visited[v]) return 0l;

    // mark node as visited 
    visited[v] = true;
    long count = 1;

    // for each u in adj. List of v call dfs and add return value to count
    for (std::pair<int, int> u : graph.adjList[v])
    {
        if (!visited[u.first])
            count += dfsCountNode(graph, visited, u.first);
    }

    return count;
}


int main()
{
    int numNodes, numEdges;
    std::cout << "Enter no. of Nodes and no. of Edges in the graph ...\n";
    std::cin >> numNodes >> numEdges;

    bool directed, weighted;
    std::cout << "Is the Graph weighted and directed (0/1) ...\n";
    std::cin >> weighted >> directed;

    int u, v, w;
    std::vector<Edge> edges;
    if (weighted)
    {
        for (int i=0; i<numEdges; i++)
        {
            std::cin >> u >> v >> w;
            edges.push_back(Edge(u, v, w));
        }
    }
    else
    {
        for (int i=0; i<numEdges; i++)
        {
            std::cin >> u >> v;
            edges.push_back(Edge(u, v));
        }
    }

    // Initializing Graph and visited array
    Graph graph = Graph(numNodes, edges, directed);
    std::vector<bool> visited1(numNodes, false);
    std::vector<bool> visited2(numNodes, false);


    long count = 0;

    for (int i=0; i<numNodes; i++)
    {
        if (!visited1[i])
            dfs(graph, visited1, i);
        if (!visited2[i])
            count += dfsCountNode(graph, visited2, i);
    }

    std::cout << "\nNumber of nodes: " << count; 
}   

File: code/python/usrIp_dfsalr.py
# global variables
input_path = '/home/freyr/dev/GraphTheory/code/python/input.txt'
output_path = '/home/freyr/dev/GraphTheory/code/python/output.txt'

# clearing the contents of the file if any
open(output_path, 'w').close()

class Edge:
    def __init__(self, src, dest, cost=1):
        self.src = src
        self.dest = dest
        self.cost = cost
        
    def __repr__(self) -> str:
        return f"Edge({self.src}, {self.dest}, {self.cost}"
    
    def __str__(self) -> str:
        return repr(self)
    

class Graph:
    def __init__(self, num_nodes, edges, directed=False):
        self.num_nodes = num_nodes
        self.adj_list = {i: [] for i in range(self.num_nodes)}
        
        for edge in edges:
            self.adj_list[edge.src].append((edge.dest, edge.cost))
            if not directed:
                self.adj_list[edge.dest].append((edge.src, edge.cost))
                
def dfs(graph, visited, node=0):
    visited[node] = True
    
    # write to file
    with open(output_path, "a") as f:
        f.write(f"{node} ")
        # print(node, end=" ")
    
    for u in graph.adj_list[node]:
        if not visited[u[0]]:
            dfs(graph, visited, u[0])
            
def dfs_count_node(graph, visited, node=0):
    if visited[node]: return 0
    
    visited[node] = True
    count = 1
    
    for u in graph.adj_list[node]:
        if not visited[u[0]]:
            count += dfs_count_node(graph, visited, u[0])
            
    return count


if __name__ == '__main__':

    
    with open(input_path, "r") as f:
    
        # print("Enter no. of Nodes and no. of Edges in the graph ...")
        num_nodes, num_edges = (int(i) for i in f.readline().split(" "))
        
        # print("Is the Graph directed and weighted (0/1) ...")
        directed, weighted = (int(i) for i in f.readline().split(" "))
        
        # print(weighted, type(weighted), directed, type(directed))
    
        edges = []
        if weighted:
            for _ in range(num_edges):
                u, v, w = (int(i) for i in f.readline().split(" "))
                edges.append(Edge(u, v, w))
        else:
            for _ in range(num_edges):
                u, v = (int(i) for i in f.readline().split(" "))
                edges.append(Edge(u, v))  
            
    graph = Graph(num_nodes, edges)
    visited1 = [False for _ in range(num_nodes)]
    visited2 = [False for _ in range(num_nodes)]
    count = 0
    
    # Output Indicator
    with open(output_path, "a") as f:
        f.write(f"\nPROGRAM OUTPUT STARTS HERE:\n\n")
    
    for i in range(num_nodes):
        if not visited1[i]:
            dfs(graph, visited1, i);
        if not visited2[i]:
            count += dfs_count_node(graph, visited2, i);
    
    with open(output_path, "a") as f:
        f.write(f"\nNumber of nodes counted by dfs: {count}\n")
    
    # print("\nNodes counted by dfs: {}".format(count))